### DISCUSSION TITLE
Debug companion app?

### DISCUSSION THREAD

**Discussion by lars18th** [termux/termux-app#2571] [Ideas]
Hi,

I'm not sure if this question is stupid. However, I prefer to ask... 😉 

Currently to execute the Termux environment from the ADB shell we need to use the "run-as" command. However, this implies that the Apk requires to be compiled as Debug. Ok, then the Termux App installed by hand requires to be in Debug. But this implies that the performance is reduced, rigth? So, why not provide a Termux-companion App, that will be compiled in Debug, and then it will work calling to the regular Termux App compiled in Release mode?

Please, if this sounds stupid or crazy, then you only need to explain the reason for it. Sorry for any noise created.
Regards.

**Comment by agnostic-apollo** [comment]
All termux github builds are debuggable and I have not noticed any performance issues under normal conditions, its still snappy. If I am not mistaken, any serious performance issues are likely gonna happen when app is connected to a debugger, like in android studio, despite any extra debug code added to the app.

If you are still worried about performance issues and want to remotely connect, use `ssh`, although that will have network issues.

https://github.com/termux/termux-app#github

https://wiki.termux.com/wiki/Remote_Access#SSH




> So, why not provide a Termux-companion App, that will be compiled in Debug, and then it will work calling to the regular Termux App compiled in Release mode?

How is the app supposed to "connect" to regular app, over which protocol?

**Reply by lars18th** [reply]
Hi @agnostic-apollo ,

Thank you for the response. The use case is this: #2565 . So the idea is to overcome the overhead of the SSH procotol, because when we use the ADB plus a wire connection we don't need encryption. And because we don't want to store passwords (even fake) or certificates to connect to the Termux shell.

However, if you say that _all_ releases of Termux are debuggable, and if this doesn't represent _any_ performance penalty; then in this case, my request doesn't have sense.

Thank you very much for your explanation.
Regards.

Note: At time I'm connecting from my PC to Termux only using this command: https://github.com/termux/termux-app/discussions/2565#discussioncomment-2116176 . I feel this mode more clean, easy and with good performance. Only I need how to enable the upper SSH protocol (multiplexation of sessions, without the lower encryption).

**Reply by agnostic-apollo** [reply]
Only github releases are debuggable, not F-Droid or Playstore. Welcome.

**Reply by lars18th** [reply]
> Only github releases are debuggable, not F-Droid or Playstore. Welcome.

Hi @agnostic-apollo ,

I know that Play Store releases (now deprecated) are not debuggable. The question is: regarding the performance, it exists some difference between debuggable/non-debuggale?

**Reply by agnostic-apollo** [reply]
How about testing it yourself with github vs f-droid builds if its that much of a concern.

**Reply by lars18th** [reply]
Hi @agnostic-apollo ,

Perhaps I'm expresed wrong. I'm using Termux with the github version. And this version is debuggable. So I can start it using the `run-as` command from the ADB shell. So this is not a problem for me now.

So, what I ask if it's possible to provide the same for a Termux _without_ debug enabled. Only that. For example, the Termux-api can call to Termux-App. The question is: if one is debuggable and the other not, but signed with the same certificate... then they can communicate? IF the answer is YES, then perhaps a Termux _non debuggable_ can be started from a _debuggable_ Termux-companion that has started with `run-as`.

I hope now the question will be more clear. Excuse me for my poor explanation before.
Regards.

**Reply by agnostic-apollo** [reply]
Define "started from a debuggable Termux-companion". And if there will be constant communication between debuggable Termux-companion and Termux non debuggable, then you would obviously have performance issues you are trying to avoid.

Secondly, `run-as` doesn't start the app process or goes through it, it just changes current uid/gid/selinux context and capabilities that would normally be assigned to the app and starts a shell process. It won't be running any debuggable code in the app, only debuggable flag is checked by `run-as`.

https://cs.android.com/android/platform/superproject/+/android-12.0.0_r4:system/core/run-as/run-as.cpp;l=164

https://cs.android.com/android/platform/superproject/+/android-12.0.0_r4:external/minijail/libminijail.c;l=2264

Any performance issues would be when app is run and likely connected to debugger. I have not noticed any slow downs, even on my 6 year old device. And I am pretty sure no termux dev would invest time to implement a companion app for any said minuscule improvements just for a few `run-as` users, even if it was possible. There are far more important things to do. And any communication protocol between the apps will obv have its own lags, so there is no point anyways.

If you have root, then you can alternatively use non-debuggable termux and start a shell in termux context. Check https://android.stackexchange.com/questions/217016/how-to-run-a-program-in-an-app-context-with-magisk. Termux `util-linux` provides `setpriv` and `nsenter` binaries now added in https://github.com/termux/termux-packages/pull/8198.

**Reply by lars18th** [reply]
Hi @agnostic-apollo ,

Thank you for the detailed explanation. Now I'm aware of some relevant information that I've missed:
- The "run-as" tool checks if the package has the "debug" flag.
- Setting the "debug" flag in the Termux has a negligible performance penalty.

In this case, I apologize for this request. It doesn't have sense. Even I've learned a lot about the Termux context. 😃 
Thank you.

**Reply by agnostic-apollo** [reply]
> negligible performance penalty.

I don't know how negligible though, that would require testing.

Welcome.
